c7n/filters/iamaccess.py

```diff
@@ -343,3 +343,116 @@ def __call__(self, r):
+                    'NotAction': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Resource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotResource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Condition': {'type': 'object'},
+                    'PartialMatch': {
+                        'anyOf': [
+                            {'type': 'string', "enum": POLICY_ELEMENTS},
+                            {'type': 'array', 'items': [
+                                {"type": "string", "enum": POLICY_ELEMENTS}]}]
+                    }
+                },
+                'required': ['Effect']
+            }
+        })
+
+    def process(self, resources, event=None):
+        return list(filter(None, map(self.process_resource, resources)))
+
+    def get_policy(self, r):
+        return r.get('Policy')
+
+    def get_required_statements(self, r):
+        return format_string_values(list(self.data.get('statements', [])))
+
+    def process_resource(self, r):
+        p = self.get_policy(r)
+        if p is None or not p:
+            return None
+        p = json.loads(p)
+
+        required = list(self.data.get('statement_ids', []))
+        statements = p.get('Statement', [])
+        for s in list(statements):
+            if s.get('Sid') in required:
+                required.remove(s['Sid'])
+
+        required_statements = self.get_required_statements(r)
+        for required_statement in required_statements:
+            partial_match_elements = required_statement.pop('PartialMatch', [])
+            if isinstance(partial_match_elements, str):
+                # Make list if string to prevent mismatches like Action in NotAction below
+                partial_match_elements = [partial_match_elements]
+            for statement in statements:
+                match_count = sum(
+                    1 for k, v in required_statement.items() if self.match(
+                        k, v, statement, partial_match_elements))
+                if match_count == len(required_statement):
+                    required_statements.remove(required_statement)
+                    break
+
+        if (self.data.get('statement_ids', []) and not required) or \
+           (self.data.get('statements', []) and not required_statements):
+            return r
+        return None
+
+    def match(self, k, v, stmt, partial_match_elements):
+        if k in stmt:
+            if k in partial_match_elements:
+                if isinstance(v, list):
+                    return set(v).issubset(stmt[k])
+                elif isinstance(v, dict):
+                    return dict_merge(stmt[k], v) == stmt[k]
+                else:
+                    return v in stmt[k]
+            return v == stmt[k]
+
+
+def dict_merge(d1, d2):
+    d3 = d1.copy()
```

Comment:
this should be in utils.py and have dedicated tests

Reply:
Good call. Moved to utils in 512fa413509e6705ac75f69fbdbe62025a18fdef. I'll add the tests this afternoon. Sorry for being awol for so long, finally cleared my plate wrt to my studies.

Reply:
Added a test and updated the existing merge dict functionality to reflect what we need in f29583e